collection_mutation.cc

```diff
@@ -107,24 +107,42 @@ bool collection_mutation_view::is_any_live(const abstract_type& type, tombstone
 }
 
 api::timestamp_type collection_mutation_view::last_update(const abstract_type& type) const {
-    assert(type.is_collection());
-    auto& ctype = static_cast<const collection_type_impl&>(type);
-
   return data.with_linearized([&] (bytes_view in) {
     api::timestamp_type max = api::missing_timestamp;
     auto has_tomb = read_simple<bool>(in);
     if (has_tomb) {
         max = std::max(max, read_simple<api::timestamp_type>(in));
         (void)read_simple<gc_clock::duration::rep>(in);
     }
+
     auto nr = read_simple<uint32_t>(in);
-    for (uint32_t i = 0; i != nr; ++i) {
-        auto ksize = read_simple<uint32_t>(in);
-        in.remove_prefix(ksize);
-        auto vsize = read_simple<uint32_t>(in);
-        auto value = atomic_cell_view::from_bytes(ctype.value_comparator()->imr_state().type_info(), read_simple_bytes(in, vsize));
-        max = std::max(value.timestamp(), max);
-    }
+
+    visit(type, make_visitor(
+        [&] (const collection_type_impl& ctype) {
+            auto& type_info = ctype.value_comparator()->imr_state().type_info();
+            for (uint32_t i = 0; i != nr; ++i) {
```

Comment:
This loop and the one for UDTs look very similar. It could make sense to generalize them and just pass lambda that reads key and value in collection/udt specific way.